### PR TITLE
add feature cosine similarity loss

### PR DESCRIPTION
**Pull request by entrpn**
# Description

Adds optional cosine similarity loss between attention outputs of various layers of teacher/student.

# Tests

- Updated train_distill_test unit test.
- Ran train_distill.py on llama3.1-8b both with cosine loss enabled/disabled.
- Updated maxtext_utils tests.

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [X] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [X] I have necessary comments in my code, particularly in hard-to-understand areas.
- [X] I have run end-to-end tests tests and provided workload links above if applicable.
- [X] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

File: src/maxtext/models/models.py [hunk 1]
```diff
@@ -471,6 +471,12 @@ def __call__(
     if audio_embeddings is not None:
       audio_masks = mm_processor.get_bidirectional_mask_audio(self.config, decoder_input_tokens)
 
+    mutable_collections = []
+    if self.config.record_internal_nn_metrics:
```
Comment (gagika) [inline src/maxtext/models/models.py:481]: should this be added into base.yml?

File: src/maxtext/models/models.py [hunk 2]
```diff
@@ -471,6 +471,12 @@ def __call__(
...
+      mutable_collections.append("intermediates")
+    if self.config.distill_beta > 0.0 and "intermediates" not in mutable_collections:
```
Comment (gagika) [inline src/maxtext/models/models.py:483]: same here for distill_beta.

File: src/maxtext/layers/attentions.py [hunk 3]
```diff
@@ -1162,5 +1162,7 @@ def __call__(
       out = out.reshape(batch_size, seq_len, self.config.num_query_heads * self.config.head_dim)
       out = out * jax.nn.sigmoid(gate)
     out = self.out_projection(out, out_sharding=out_sharding)
+    if self.config.distill_beta > 0.0:
+      self.sow(nnx.Intermediate, "distill_features", out)
```
Comment (gagika) [inline src/maxtext/layers/attentions.py]: will something like this also work for NNX?

self.distill_features = nnx.Intermediate(out)

also should you Raise ValueError in src/maxtext/configs/types.py distill_beta > 0.0 and model is run in Linen?

File: src/maxtext/trainers/post_train/distillation/distillation_utils.py [hunk 4]
```diff
@@ -110,9 +110,52 @@ def __next__(self) -> MaxTextTrainingInput:
...
+      teacher_forward_fn: Callable[..., jax.Array],
+      labels_fn: Callable[..., jax.Array],
+      temperature: float = 2.0,
+      alpha: float = 0.5,
+      beta_feature: float = 0.0,
+      layer_indices: Optional[List[int]] = None,
+      feature_loss_fn: Callable[[jax.Array, jax.Array], jax.Array] | None = None,
+      cosine_distance_axis: int | tuple[int, ...] = -1,
+  ):
+    """Initializes the Logit strategy using tunix logit.LogitStrategy.
```
Comment (gagika) [inline src/maxtext/trainers/post_train/distillation/distillation_utils.py]: Initializes the CombinedDistillationStrategy ...

File: src/maxtext/trainers/post_train/distillation/train_distill.py [hunk 5]
```diff
@@ -135,14 +135,18 @@ def model_forward_fn(model, input_tokens, positions, attention_mask, decoder_seg
...
     logits = model(
         decoder_input_tokens=input_tokens,
         decoder_positions=positions,
         decoder_segment_ids=decoder_segment_ids,
         enable_dropout=config.enable_dropout,
     )
-    return logits
+    hidden_features = None
+    if config.distill_beta > 0.0:
+      hidden_features = model.decoder.layers["self_attention"]["distill_features"].get_value()[-1]
```
Comment (gagika) [inline src/maxtext/trainers/post_train/distillation/train_distill.py]: Replace both with:

hidden_features = maxtext_utils.get_intermediate_value(model, "distill_features", clear=True) ?

File: src/maxtext/models/models.py (hunk 1 above)
Comment (entrpn) [reply src/maxtext/models/models.py:481]: it was already a config in base.yml. I included it here since it was never used with nnx.

File: src/maxtext/trainers/post_train/distillation/train_distill.py (hunk 5 above)
Comment (entrpn) [reply src/maxtext/trainers/post_train/distillation/train_distill.py]: done

File: src/maxtext/layers/attentions.py (hunk 3 above)
Comment (entrpn) [reply src/maxtext/layers/attentions.py]: no, this will just set the distill_features value to an nnx.Intermediate but it won't append them like sow does.

Comment (entrpn) [reply src/maxtext/layers/attentions.py]: I don't see a way to check if the model is run in Linen. There is a variable `enable_nnx` in the config but even with this false, we will still run with the nnx linen bridge. What about just letting flax handle the failure?

File: src/maxtext/configs/types.py [hunk 6]
```diff
@@ -1051,6 +1052,8 @@ class Distillation(BaseModel):
   # --- Loss Params ---
   distill_alpha: float = Field(0.5, description="Weight for the distillation loss component.")
   distill_temperature: float = Field(1.0, description="Temperature for distillation softening.")
+  distill_beta: float = Field(0.0, description="Weight for the feature loss component.")
+  distill_layer_indices: list = Field([0, 1, 2, 3, 4, 5, 6, 7], description="Feature indices for feature loss.")
```
Comment (gagika) [inline src/maxtext/configs/types.py]: those need to be added in base.yml

File: src/maxtext/trainers/post_train/distillation/distillation_utils.py [hunk 7]
```diff
@@ -110,9 +110,52 @@ def __next__(self) -> MaxTextTrainingInput:
...
+        teacher_forward_fn: Inherited from `logit.LogitStrategy`. Function to compute teacher model outputs.
+        labels_fn: Inherited from `logit.LogitStrategy`. Function to compute labels from model inputs.
+        temperature: Inherited from `logit.LogitStrategy`. Temperature for softening probabilities (> 0).
+        alpha: Inherited from `logit.LogitStrategy`. Weight to balance distillation loss and task loss (0.0 to 1.0).
+        beta_feature: Weight to balance feature loss (0.0 to 1.0). 0.0 disables feature loss.
+        layer_indices: Layer indices to apply feature loss.
+        feature_loss_fn: A function that takes two jax. Arrays (student_map,
+          teacher_map) and returns a scalar loss. Defaults to Cosine Distance.
+        cosine_distance_axis: The axis to use for cosine distance computation if
+          feature_loss_fn is not provided. Defaults to -1.
```
Comment (gagika) [inline src/maxtext/trainers/post_train/distillation/distillation_utils.py:141]: `Defaults to -1.` -> cosine_distance loss ?

File: src/maxtext/trainers/post_train/distillation/distillation_utils.py [hunk 8]
```diff
@@ -144,14 +197,31 @@ def compute_loss(
...
+
+    total_loss = base_logit_loss + feature_loss
 
     # 4. Return Loss AND Metrics
     metrics = {
         "distill/soft_loss": soft_loss,
         "distill/hard_loss": hard_loss,
         "distill/kl_div": jnp.mean(kl_div),
         "distill/teacher_loss": teacher_hard_loss,
+        "distill/feature_loss": feature_loss,
```
Comment (gagika) [inline src/maxtext/trainers/post_train/distillation/distillation_utils.py]: is `feature_loss` common name? wondering if `cosine_similarity_loss` or `activation_loss` is more descriptive?

**Review by JamesDeng42 — APPROVED** [review]

File: src/maxtext/trainers/post_train/distillation/distillation_utils.py [hunk 9]
```diff
@@ -123,8 +166,18 @@ def compute_loss(
     # Calculate Distillation Loss (KL Divergence)
     # Scale logits by temperature T for soft targets
     # We use explicit float32 casting for stability in loss calculation
-    s_logits = student_output.astype(jnp.float32)
-    t_logits = teacher_output.astype(jnp.float32)
+    s_logits = student_output[0].astype(jnp.float32)
```
Comment (vlad-karp) [inline src/maxtext/trainers/post_train/distillation/distillation_utils.py:169]: change the type of student_output & teacher_output in the function signature. It is a tuple nnow

File: src/maxtext/trainers/post_train/distillation/distillation_utils.py [hunk 10]
```diff
@@ -144,14 +197,31 @@ def compute_loss(
...
+        # jnp.take slices along axis=0 (the layer dimension)
+        s_features_sliced = jnp.take(s_features, self.layer_indices, axis=0)
+        t_features_sliced = jnp.take(t_features, self.layer_indices, axis=0)
+      else:
+        s_features_sliced = s_features
+        t_features_sliced = t_features
+
+      feature_loss = self.beta_feature * self.feature_loss_fn(s_features_sliced, t_features_sliced)
+
+    total_loss = base_logit_loss + feature_loss
```
Comment (vlad-karp) [inline src/maxtext/trainers/post_train/distillation/distillation_utils.py:215]: should we use some weighting factor here as well when combining losses?

**Review by vlad-karp — COMMENTED** [review]
A couple of comments

File: src/maxtext/trainers/post_train/distillation/distillation_utils.py (hunk 10 above)
Comment (entrpn) [reply src/maxtext/trainers/post_train/distillation/distillation_utils.py:215]: there are two weight factors, alpha and beta_feature. This is just the final total loss being added together.

File: src/maxtext/trainers/post_train/distillation/distillation_utils.py (hunk 9 above)
Comment (entrpn) [reply src/maxtext/trainers/post_train/distillation/distillation_utils.py:169]: I will address this in a follow-up PR

File: src/maxtext/trainers/post_train/distillation/distillation_utils.py (hunk 10 above)
Comment (vlad-karp) [reply src/maxtext/trainers/post_train/distillation/distillation_utils.py:215]: ok, I see, you use self.beta_feature as a non-normalized loss factor. I assume we wanted to use a normalized form when combining.

File: src/maxtext/utils/maxtext_utils.py [hunk 11]
```diff
@@ -897,6 +897,34 @@ def get_nested_value(dictionary, nested_key, default=None):
...
+    clear: Clears the intermediate value from the model.
+
+  Returns:
+    The value associated with the nested key, or the default value if not found.
+  """
+  intermediate_value = default
+  match nested_key:
+    case "out_projection_activations":
+      if nested_key in model.decoder.layers["self_attention"]:
+        intermediate_value = model.decoder.layers["self_attention"][nested_key].get_value()[-1]
```
Comment (gagika) [inline src/maxtext/utils/maxtext_utils.py:918]: why is it only returning the last element e.g. [-1]?
what is the shape of model.decoder.layers["self_attention"][nested_key] ? wondering what is getting dropped?

File: src/maxtext/trainers/post_train/distillation/distillation_utils.py
Comment (gagika) [inline src/maxtext/trainers/post_train/distillation/distillation_utils.py:236]: should you do `s_logits = student_output[0].astype(jnp.float32)` similar to compute_loss now that model_forward_fn returns a tuple? also we can have a TODO to add other metrics in eval.

looks like our tests are not testing this function?

File: src/maxtext/utils/maxtext_utils.py (hunk 11 above)
Comment (entrpn) [reply src/maxtext/utils/maxtext_utils.py:918]: this is because sow appends values in a tuple, so its just a way to retrieve it.

File: src/maxtext/trainers/post_train/distillation/distillation_utils.py
Comment (entrpn) [reply src/maxtext/trainers/post_train/distillation/distillation_utils.py:236]: yes this will cause eval to fail. I'll fix this and add unit tests.

File: src/maxtext/utils/maxtext_utils.py (hunk 11 above)
Comment (gagika) [reply src/maxtext/utils/maxtext_utils.py:918]: in a follow up PR could you comment what's inside the tuple and what is being retrieved here?
